Product Backlog Item 2124976: SaaS App Tutorial: Graph APIs to configure SAML-based single sign-on Update (#18890)
microsoftgraph/microsoft-graph-docs: concepts/application-saml-sso-configure-api.md

}
```

> ### Create a custom signing certificate

Reviewer: ```suggestion
### Option 2: Create a custom signing certificate
```

            using (RSA rsa = RSA.Create(2048))
            {
                var request = new CertificateRequest(distinguishedName, rsa, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);

                request.CertificateExtensions.Add(
                    new X509KeyUsageExtension(X509KeyUsageFlags.DataEncipherment | X509KeyUsageFlags.KeyEncipherment | X509KeyUsageFlags.DigitalSignature, false));

                var certificate = request.CreateSelfSigned(new DateTimeOffset(certificateStartDate), new DateTimeOffset(certificateEndDate));
                certificate.FriendlyName = CertificateName;

                return new X509Certificate2(certificate.Export(X509ContentType.Pfx, password), password, X509KeyStorageFlags.Exportable);
            }
        }

        // Generate hash from thumbprint.
        public static string GetSha256FromThumbprint(string thumbprint)
        {
            var message = Encoding.ASCII.GetBytes(thumbprint);
            SHA256Managed hashString = new SHA256Managed();
            return Convert.ToBase64String(hashString.ComputeHash(message));
        }
    }
}
```

> ### Add a custom signing key

Reviewer: You've confirmed internally that this is required step when you use the C# code
```suggestion
#### Add a custom signing key
```

> Add the following information to the service principal:

Reviewer: ```suggestion
After running the previous code, extract the values for the following parameters from the PFX file. You will add this information to the service principal:
```

* Private key
* Password
* Public key 

> Extract the private and public key Base64 encoded from the PFX file. To learn more about the properties, see [keyCredential resource type](/graph/api/resources/keycredential?view=graph-rest-1.0&preserve-view=true).

Reviewer: ```suggestion
For more information about the properties, see [keyCredential resource type](/graph/api/resources/keycredential).
```

Make sure that the keyId for the keyCredential used for "Sign" matches the keyId of the passwordCredential. You can generate the `customkeyIdentifier` by getting the hash of the cert's thumbprint. See the previous C# reference code.

> #### Request

Reviewer: ```suggestion
##### Request
```

> [!NOTE]
> The "key" value in the keyCredentials property is shortened for readability. The value is base 64 encoded. For the private key the property `usage` is "Sign". For the public key the property `usage` is "Verify".

> <!-- {
>   "blockType": "request",
>   "name": "servicePrincipals"
> }-->
> ```msgraph-interactive
> PATCH https://graph.microsoft.com/v1.0/servicePrincipals/f47a6776-bca7-4f2e-bc6c-eec59d058e3e
> 
> Content-type: application/json
> 
> {

Reviewer: ```suggestion
<!-- {
  "blockType": "request",
  "name": "tutorial_configure_saml_sso_add_credentials"
}-->
```http
PATCH https://graph.microsoft.com/v1.0/servicePrincipals/f47a6776-bca7-4f2e-bc6c-eec59d058e3e
Content-type: application/json

{
```

            "endDateTime": "2021-04-22T22:10:13Z",
            "keyId": "4c266507-3e74-4b91-aeba-18a25b450f6e",
            "startDateTime": "2020-04-22T21:50:13Z",
            "type": "X509CertAndPassword",
            "usage": "Sign",
            "key":"MIIKIAIBAz.....HBgUrDgMCERE20nuTptI9MEFCh2Ih2jaaLZBZGeZBRFVNXeZmAAgIH0A==",
            "displayName": "CN=awsAPI"
        },
        {
            "customKeyIdentifier": "lY85bR8r6yWTW6jnciNEONwlVhDyiQjdVLgPDnkI5mA=",
            "endDateTime": "2021-04-22T22:10:13Z",
            "keyId": "e35a7d11-fef0-49ad-9f3e-aacbe0a42c42",
            "startDateTime": "2020-04-22T21:50:13Z",
            "type": "AsymmetricX509Cert",
            "usage": "Verify",
            "key": "MIIDJzCCAg+gAw......CTxQvJ/zN3bafeesMSueR83hlCSyg==",
            "displayName": "CN=awsAPI"
        }

    ],
    "passwordCredentials": [
        {
            "customKeyIdentifier": "lY85bR8r6yWTW6jnciNEONwlVhDyiQjdVLgPDnkI5mA=",
            "keyId": "4c266507-3e74-4b91-aeba-18a25b450f6e",
            "endDateTime": "2022-01-27T19:40:33Z",
            "startDateTime": "2020-04-20T19:40:33Z",
            "secretText": "61891f4ee44d"
        }
    ]
}
```

> #### Response
> 
> <!-- {
>   "blockType": "response",
>   "truncated": true,
> } -->

Reviewer: ```suggestion
##### Response

<!-- {
  "blockType": "response",
  "truncated": true
} -->
```
